### PR TITLE
Ensure lazy evaluation for probs and logits

### PR DESCRIPTION
**Pull request by neerajprad**
This makes a few modifications to ensure that `logits` and `probs` are not accessed by a method unless it needs to. e.g. accessing `probs` on a class that was initialized with `logits` only to get the batch shape or event shape. This affects Categorical, OneHotCategorical, Bernoulli and Multinomial.

**Testing:** Added some tests to prevent future regression:
 - `initializing` a distribution with `probs` and calling `sample`, `enumerate_support`, `batch_shape` and `event_shape` should not initialize `logits`.
 - Conversely, `initializing` a distribution with `logits` and calling `log_prob`, `enumerate_support`, `batch_shape` and `event_shape` should not initialize `probs`.

### CODE REVIEW

**Comment by neerajprad** [comment]
Closing this. I intended for this to be in probtorch first.
